File: Assignment3_NQueens.py
# Assignment 3: N-Queens Problem using Branch and Bound / Backtracking
# Implement a solution for a Constraint Satisfaction Problem
# using Branch and Bound and Backtracking for n-queens problem

def is_safe(board, row, col, n):
    """Check if placing queen at (row, col) is safe"""
    # Check column
    for i in range(row):
        if board[i] == col:
            return False
    
    # Check upper-left diagonal
    for i, j in zip(range(row-1, -1, -1), range(col-1, -1, -1)):
        if board[i] == j:
            return False
    
    # Check upper-right diagonal
    for i, j in zip(range(row-1, -1, -1), range(col+1, n)):
        if board[i] == j:
            return False
    
    return True

def solve_nqueens(board, row, n):
    """Solve N-Queens using backtracking"""
    if row == n:
        return True  # All queens placed
    
    for col in range(n):
        if is_safe(board, row, col, n):
            board[row] = col  # Place queen
            if solve_nqueens(board, row + 1, n):
                return True
            board[row] = -1  # Backtrack
    
    return False

def print_board(board, n):
    """Print the chessboard"""
    for i in range(n):
        row = ""
        for j in range(n):
            if board[i] == j:
                row += " Q "
            else:
                row += " . "
        print(row)
    print()

# --- Main ---
if __name__ == "__main__":
    n = int(input("Enter number of queens (N): "))
    
    # board[i] = column position of queen in row i
    board = [-1] * n
    
    print(f"\n=== {n}-Queens Problem using Backtracking ===\n")
    
    if solve_nqueens(board, 0, n):
        print(f"Solution found for {n}-Queens:\n")
        print_board(board, n)
        print("Queen positions (row, col):")
        for i in range(n):
            print(f"  Row {i+1}: Column {board[i]+1}")
    else:
        print(f"No solution exists for {n}-Queens!")

# Sample: N=4
# Output:
#  .  Q  .  .
#  .  .  .  Q
#  Q  .  .  .
#  .  .  Q  .

File: Assignment3_NQueens.cpp
// Assignment 3: N-Queens Problem using Branch and Bound / Backtracking
// Implement a solution for a Constraint Satisfaction Problem
// using Branch and Bound and Backtracking for n-queens problem
// Linux/Ubuntu: g++ -std=c++17 Assignment3_NQueens.cpp -o Assignment3_NQueens && ./Assignment3_NQueens
// Fedora: sudo dnf install gcc-c++ && g++ -std=c++17 Assignment3_NQueens.cpp -o Assignment3_NQueens && ./Assignment3_NQueens

#include <iostream>
using namespace std;

int board[20];  // board[i] = column of queen in row i

// Check if placing queen at (row, col) is safe
bool isSafe(int row, int col, int n) {
    for (int i = 0; i < row; i++) {
        // Same column
        if (board[i] == col) return false;
        // Same diagonal
        if (abs(board[i] - col) == abs(i - row)) return false;
    }
    return true;
}

// Solve using backtracking
bool solve(int row, int n) {
    if (row == n) return true;  // All queens placed

    for (int col = 0; col < n; col++) {
        if (isSafe(row, col, n)) {
            board[row] = col;           // Place queen
            if (solve(row + 1, n))
                return true;
            board[row] = -1;            // Backtrack
        }
    }
    return false;
}

// Print the board
void printBoard(int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (board[i] == j)
                cout << " Q ";
            else
                cout << " . ";
        }
        cout << endl;
    }
    cout << endl;
}

int main() {
    int n;
    cout << "Enter number of queens (N): ";
    cin >> n;

    // Initialize board
    for (int i = 0; i < n; i++) board[i] = -1;

    cout << "\n=== " << n << "-Queens Problem using Backtracking ===" << endl << endl;

    if (solve(0, n)) {
        cout << "Solution found for " << n << "-Queens:" << endl << endl;
        printBoard(n);
        cout << "Queen positions (row, col):" << endl;
        for (int i = 0; i < n; i++)
            cout << "  Row " << i + 1 << ": Column " << board[i] + 1 << endl;
    } else {
        cout << "No solution exists for " << n << "-Queens!" << endl;
    }

    return 0;
}

// Sample Input:
// 4
// Sample Output:
//  .  Q  .  .
//  .  .  .  Q
//  Q  .  .  .
//  .  .  Q  .
